Seeshark 5

Speykious opened the pull request:
For SeeShark 5, I chose to completely rewrite the entire thing.

Having to deal with FFmpeg setup is becoming tiring. Some missing features meant that I had to write interop code for each OS anyway, and since I'm always tied to a specific version of it, it's also not very flexible.
So I decided it's best to not use FFmpeg at all, and to instead use the OS APIs directly. That way SeeShark even has a chance to be somewhat *done* at some point in the future. At the very least, it'll be easier to maintain, with no external dependencies, and the only things to care about being OS API changes and .NET version updates.

As a side note, since this is where SeeShark is going, I might decide to add support for retrieving audio in the future. There's something to be done both for cameras and displays.

# Features to be implemented

Meaning of the symbols:
- &check; — implemented
- &cross; — not implemented
- &not; — incompatible

## Camera

| Feature | Linux<br>(V4L2) | MacOS<br>(AVFoundation) | Windows<br>(DirectShow) |
|---|---|---|---|
| Get available cameras | &check; | &check; | &check; |
| Get available formats of camera | &check; | &check; | &check; |
| Open camera | &check; |&check; | &check; |
| Open camera with video format options | &check; | &check; | &cross; |
| Configure camera with video format options | &check; | &check; | &cross; |
| Capture raw frames from camera | &check; | &check; | &cross; |
| Notify (dis)connections of cameras | &cross; | &cross; | &cross; |

## Display

| Feature | Linux<br>(X11) | Linux<br>(Wayland) | MacOS<br>(ScreenCaptureKit) | Windows<br>(GDI) |
|---|---|---|---|---|
| Get available displays | &cross; | &cross; | &cross; | &cross; |
| Get available formats of a display | &cross; | &cross; | &cross; | &cross; |
| Open display | &cross; | &cross; | &cross; | &cross; |
| Open display with video format options | &cross; | &cross; | &cross; | &cross; |
| Capture raw frames from display | &cross; | &cross; | &cross; | &cross; |

# Status of the PR

Currently implementing camera support for Windows.
I initially went with DShow, but I think I'm gonna scrap all of it and use the newer MediaFoundation API instead.

Speykious committed:
The overhaul begins.

Speykious committed:
Change license to BSD 2-Clause

Speykious committed:
WIP Readme edit

Speykious committed:
Refactoring, SupportedOSPlatform, video format

Speykious committed:
Don't use external terminal for debugging (yet)

Speykious committed:
Fixes, use libc select to read frame

Speykious committed:
Update file header (license)

Speykious committed:
It's a surprise tool that will help us later (・u・/)

Speykious committed:
Make all interop code internal

Speykious committed:
Sponsorship button because why not

Somfic commented:
If you need help testing on MacOS/Windows let me know

adryzz commented:
as for GDI+, i have enough knowledge of the internals that i can make an implementation of display capture fairly quickly, but currently i'm... *stuck* in uni so it'll take a few days before i start, i think

Speykious committed:
Basic MacOS interop

Speykious commented:
Progress!
![image](https://github.com/Speykious/SeeShark/assets/13885008/b8ef58a8-c780-4df9-aa15-f0f59337d43b)

Speykious committed:
?

Speykious committed:
Start AVFoundation methods impl

Speykious committed:
The dictionary was REALLY bothering me

Signed-off-by: Speykious <speykious@gmail.com>

Speykious committed:
Get framerate ranges and dimensions

Signed-off-by: Speykious <speykious@gmail.com>

Speykious committed:
Implement this function at 4am because why not

Signed-off-by: Speykious <speykious@gmail.com>

Speykious committed:
Read pixels from sample buffer

Signed-off-by: Speykious <speykious@gmail.com>

Speykious commented:
After almost a full year of neglect on this PR, I came back to it yesterday and now I have a feed :3

https://github.com/user-attachments/assets/90f61400-3735-4040-8274-5a01b131b54c

Obtained by writing the raw pixel buffers to a file and then making `ffmpeg` read that using `-f raw video`, in my case:
```sh
ffmpeg -f rawvideo -video_size 1280x720 -pixel_format uyvy422 -i camera-feed.1280x720.yuv2.raw camera-feed.mp4
```

Something interesting about the pixel format of my camera is that the FourCC code given by Apple is `yuv2`, which I assumed to be equivalent to FFmpeg's `yuv422p` pixel format but nope it's not, so I tried other pixel formats that were 16 bits wide and had "422" in the name until I landed on the right one.

Speykious committed:
Get FourCC of V4l2 pixel formats (?)

Signed-off-by: Speykious <speykious@gmail.com>

Speykious committed:
V4L2 has a specific strategy for FourCC codes of BE pixel formats

Signed-off-by: Speykious <speykious@gmail.com>

Speykious committed:
Implement CVPixelFormatType

Also added a miscellaneous script used to generate this thing.

Signed-off-by: Speykious <speykious@gmail.com>

Speykious committed:
Get Apple pixel format definitions directly from origin ObjC source header (CVPixelBuffer.h)

That script was unnecessary.

Signed-off-by: Speykious <speykious@gmail.com>

Speykious committed:
Refine pixel format code

Signed-off-by: Speykious <speykious@gmail.com>

Speykious commented:
I knew I was gonna have to deal with tons of pixel formats, but it seems like most of them use some kind of FourCC code with variations. Looking around, the documentation on this stuff is very sparse and incomplete.
So as a note to self I'll keep note of a few sources:
- [FourCC](https://wiki.multimedia.cx/index.php/FourCC)
- [fourcc.org](https://fourcc.org/)
- [Video FourCCs](https://wiki.multimedia.cx/index.php/Category:Video_FourCCs)
- OS sources
  - [Windows Video FourCCs](https://learn.microsoft.com/en-gb/windows/win32/medfound/video-fourccs)
  - [Apple Pixel Format Identifiers](https://developer.apple.com/documentation/corevideo/pixel-format-identifiers?language=objc)
    - `CVPixelBuffer.h` header file containing all definitions with their corresponding FourCC if it exists, available on MacOS at `/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/System/Library/Frameworks/CoreVideo.framework/Versions/A/Headers/CVPixelBuffer.h`
  - [Linux Pixel Formats](https://www.kernel.org/doc/html/v4.10/media/uapi/v4l/pixfmt.html)
    - [`videodev2.h`](https://github.com/torvalds/linux/blob/7cdabafc001202de9984f22c973305f424e0a8b7/include/uapi/linux/videodev2.h#L541-L831) header file containing all definitions with their corresponding FourCC (they all use a FourCC in this case)

Speykious committed:
Generic ImageFormat, abstract CameraDevice

Signed-off-by: Speykious <speykious@gmail.com>

Speykious committed:
Implement CameraDevice.(Available|AvailableFormats) for MacOS

Signed-off-by: Speykious <speykious@gmail.com>

Speykious committed:
MacOS: implement request for camera permission

Signed-off-by: Speykious <speykious@gmail.com>

Speykious commented:
Permissions!

<img width="418" alt="image" src="https://github.com/user-attachments/assets/582a24fb-76c5-4d48-b7e2-40284c234bb4" />

Speykious committed:
MacOS: implement MacOSCameraDevice

There is still a lot of stuff to change.

Signed-off-by: Speykious <speykious@gmail.com>

Speykious commented:
I finally have a "functional" `MacOSCameraDevice`.

There's one big issue that still remains though: controlling the video format.
Throughout my experimentations, I found that:
- It completely ignores when I try to manually set the video dimensions, and resets it to whatever it wants
- There is a preset setting that you can set on the `AVCaptureSession` (`AVCaptureSessionPreset...`) and *that* controls the dimensions of the video output
- There is a `AVCaptureSessionPresetInputPriority` for manual configuration, except that it's the *only one I cannot use* for some reason, meaning I just can't set the video dimensions manually
- The default preset is `AVCaptureSessionPresetHigh`. When I try `AVCaptureSessionPresetMedium` instead, it gives me dimensions that aren't actually the dimensions of the camera, but of a *fictional square* that MacOS then uses to *fit* the camera into it:
  ```
  NSDictionary
  {
    AVVideoScalingModeKey               : string "AVVideoScalingModeFit"
    PixelFormatType                     : num (846624121)
    Height                              : num (480)
    Width                               : num (480)
  }
  ```

I'll likely have to find a way to programmatically get the `AVCaptureDeviceFormat` that's actually gonna be used using this fictional square.

Speykious commented:
Did a bunch more testing today; I finally managed to set both the video dimensions *and* pixel format manually without using a preset. I don't know why it was completely ignoring them before, but I don't exactly remember what I did to make that happen at that time, and it's not doing that anymore now, so I'm probably doing something right.

I also experimented with the `AVVideoScalingMode`s (see [`AVVideoScalingModeKey`](https://developer.apple.com/documentation/avfoundation/avvideoscalingmodekey?language=objc)). I'm somewhat lucky that MacOS can handle resizing on its own.

Something I discovered related to that though (with some help of a friend of mine), is that when you give a custom dimension to the `AVCaptureVideoDataOutput` instance's `videoSettings`, it will actually give you back a buffer that corresponds to an image that has the same height, but a bigger width, specifically the closest multiple of 16 from the specified width. So if I specify dimensions 659*581, it will actually give me a buffer of 672*581. And if you take a look at that, it'll have a border on the right. Maybe it's a bug that I have to forward upstream...

Either way that's now a fixed problem! I'm *this* close to finishing the MacOS camera backend.

https://github.com/user-attachments/assets/902478c6-2b33-4d1a-990c-e6a34e68fff6

Speykious committed:
MacOS: Implement video format configuration

Notably video dimensions and pixel format.
Framerate is still not configurable right now, something to investigate later.

Signed-off-by: Speykious <speykious@speykious.dev>

Speykious committed:
Fix license, readme and csproj

Signed-off-by: Speykious <speykious@speykious.dev>

Speykious committed:
MacOS: Implement framerate configuration

Signed-off-by: Speykious <speykious@speykious.dev>

Speykious committed:
MacOS: Include pixel format in video format, polish configuration code, better error handling

Signed-off-by: Speykious <speykious@speykious.dev>

Speykious committed:
Test code: use ffmpeg to create a video out of the frames when available

Signed-off-by: Speykious <speykious@speykious.dev>

Speykious committed:
MacOS: Properly handle annoying framerates

Signed-off-by: Speykious <speykious@speykious.dev>

Speykious committed:
Linux: Move interop code into an Interop/Linux folder

Signed-off-by: Speykious <speykious@speykious.dev>

Speykious committed:
MacOS: Add missing SupportedOSPlatform attributes

Speykious committed:
Finally make SeeShark a Library, have example csproj, consolidate ImageFormat, retrieve name from V4L2 devices

Not the most organized of commits I must say

Speykious committed:
MacOS: Fix wrong UYVY bit size

Signed-off-by: Speykious <speykious@speykious.dev>

Speykious committed:
Linux: Use CameraDeviceIOException, fix comments

Speykious committed:
Linux: Set frame size based on video options

Speykious committed:
Linux: Move camera open function into Linux camera device constructor

Speykious committed:
Refactor ImageFormat using some code from SeeShark 4

I still don't quite know what I'm doing with this one.

Speykious committed:
Windows: Start implementation

Speykious committed:
Linux: fields didn't need to be `internal`

Speykious committed:
Windows: Now half-working

Speykious committed:
MediaSubType.ARGB32 is actually BGRA

Speykious committed:
Update Readme

It's been there long enough.

Speykious committed:
Merge branch 'main' into seeshark-5
